numba/tests/test_parfors.py

```diff
@@ -1957,6 +1957,93 @@ def test_impl(x, sz):
             return x
         self.check(test_impl, np.zeros((10, 10)), 3)
 
+    def test_prange_unknown_call1(self):
+        @register_jitable
+        def issue7854_proc(u, i, even, size):
+            for j in range((even + i + 1) % 2 + 1, size - 1, 2):
+                u[i, j] = u[i + 1, j] + 1
+
+        # issue7854
+        # Forbid fusion in unanalyzable call inside prange.
+        def test_impl(size):
+            u = np.zeros((size, size))
+            for i in numba.prange(1, size - 1):
+                issue7854_proc(u, i, 0, size)
+            for i in numba.prange(1, size - 1):
+                issue7854_proc(u, i, 1, size)
+            return u
+
+        self.assertEqual(countParfors(test_impl, (types.int64,)), 3)
+        self.check(test_impl, 4)
+
+    def test_prange_index_calc1(self):
+        # Should forbid fusion due to cross-iteration dependency as
+        # detected by loop index calcuation (i+1) as array index.
+        def test_impl(size):
+            u = np.zeros((size, size))
+            for i in numba.prange(1, size - 1):
+                for j in range((i + 1) % 2 + 1, size - 1, 2):
+                    u[i, j] = u[i + 1, j] + 1
+            for i in numba.prange(1, size - 1):
+                for j in range(i % 2 + 1, size - 1, 2):
+                    u[i, j] = u[i + 1, j] + 1
+            return u
+
+        self.assertEqual(countParfors(test_impl, (types.int64,)), 3)
+        self.check(test_impl, 4)
+
+    def test_prange_reverse_order1(self):
+        # Testing if reversed loop index usage as array index
+        # prevents fusion.
+        def test_impl():
+            size = 10
+            a = np.zeros((size, size))
+            b = np.zeros((size, size))
+            for i in numba.prange(size):
+                for j in range(size):
+                    a[i,j] = b[i, j] + 1
+            for i in numba.prange(size):
+                for j in range(size):
+                    b[j,i] = 3
+            return a
+
+        self.assertEqual(countParfors(test_impl, ()), 3)
+        self.check(test_impl)
+
+    def test_prange_parfor_index_then_not(self):
+        # Testing if accessing an array first with a parfor index then
+        # without will prevent fusion.
+        def test_impl():
+            size = 10
+            a = np.zeros(size)
+            for i in numba.prange(size):
+                a[i] = i
+            b = 0
+            for i in numba.prange(size):
+                b += a[5]
+            return b
+
+        self.assertEqual(countParfors(test_impl, ()), 2)
```

Comment:
Just a thought... is counting the number of parfors here sufficient to guarantee that the fusion alg worked as expected. Whilst there should be 2 parfors present, there's nothing enforcing that they occurred in the way that was expected, e.g. maybe the fusion alg is wrong, and the two explicit `prange` loops have fused, and then refused to fuse with the `zeros`, opposed to the intended `zeros`+first `prange` and second `prange` combination. Perhaps the `TestParforsDiagnostics.assert_fusion_equivalence()` (and associated utilities) in this module might help?

Reply:
I rewrote this test and the next so that there's only two parfors that should fuse so there's no possible confusion.

Reply:
Thanks, I think the other tests added in this PR need similar?